lib/sequent/core/persistors/replay_optimized_postgres_persistor.rb

```diff
@@ -287,22 +287,18 @@ def do_with_record(record_class, where_clause)
         end
 
         def find_records(record_class, where_clause)
-          if @record_index.use_index?(record_class, where_clause)
-            @record_index.find(record_class, where_clause)
-          else
-            @record_store[record_class].select do |record|
-              where_clause.all? do |k, v|
-                expected_value = v.is_a?(Symbol) ? v.to_s : v
-                actual_value = record[k.to_sym]
-                actual_value = actual_value.to_s if actual_value.is_a? Symbol
-                if expected_value.is_a?(Array)
-                  expected_value.include?(actual_value)
-                else
-                  actual_value == expected_value
-                end
+          (@record_index.find(record_class, where_clause) || @record_store[record_class].select do |record|
+            where_clause.all? do |k, v|
+              expected_value = v.is_a?(Symbol) ? v.to_s : v
+              actual_value = record[k.to_sym]
+              actual_value = actual_value.to_s if actual_value.is_a? Symbol
+              if expected_value.is_a?(Array)
+                expected_value.include?(actual_value)
+              else
+                actual_value == expected_value
               end
             end
-          end.dup
+          end).dup
```

Comment:
I think this `dup` only needs to happen when the records are found in an index. So I propose moving this to `Index#find`.

Also, can it just be `freeze` and let the caller `dup` if required? Or is this too much of a trap?

Reply:
I find this `dup` a bit weird indeed. I can't really recall why this is needed, and can't find anything in old commit messages unfortunately. It has been in there since the beginning. Since it is only the array that is dupped, not the objects in it, I am not sure what it supposed to protect against....